### ISSUE TITLE
potential for better type inference with Vararg

### ISSUE THREAD

**Issue by simeonschaub** [JuliaLang/julia#37316]
Since for example `Vararg{Any,Int}` throws an error during construction, would it be possible to take advantage of the fact that the `N` parameter in a `Vararg` is always an `Int` in type inference? It would be nice if the following inferred more precisely:

```julia
julia> f(::NTuple{N}) where {N} = N
f (generic function with 1 method)

julia> code_warntype(f, Tuple{Tuple})
Variables
  #self#::Core.Compiler.Const(f, false)
  #unused#::Tuple{Vararg{T,N}} where T where N

Body::Any
1 ─     return $(Expr(:static_parameter, 1))
```
Sorry if there is an issue about this already, but I couldn't find one.

**Comment by JeffBezanson** [comment]
Yes, it looks like we should add this.

**Comment by ranocha** [comment]
> Closes ... #37316

It looks like this should have been closed by https://github.com/JuliaLang/julia/pull/38136?

**Comment by simeonschaub** [comment]
Looks like we still don't infer this on nightly:
```julia
julia> f(::NTuple{N}) where {N} = N
f (generic function with 1 method)

julia> code_warntype(f, Tuple{Tuple})
Variables
  #self#::Core.Const(f)
  #unused#::Tuple{Vararg{T, N}} where T where N

Body::Any
1 ─     return $(Expr(:static_parameter, 1))
```

**Comment by simeonschaub** [comment]
Ah, but it does infer now if the Tuple has a concrete eltype:
```julia
julia> f(::NTuple{N}) where {N} = N
f (generic function with 1 method)

julia> code_warntype(f, Tuple{Tuple{Vararg{Int}}})
Variables
  #self#::Core.Const(f)
  #unused#::Tuple{Vararg{Int64}}

Body::Int64
1 ─     return $(Expr(:static_parameter, 1))
```

**Comment by aviatesk** [comment]
Fixed by #51449.
